sycl/include/sycl/detail/named_swizzles_mixin.hpp

```diff
@@ -0,0 +1,823 @@
+  template <int N = NumElements, typename Self_ = Self>                        \
+  std::enable_if_t<                                                            \
+      (COND), decltype(std::declval<Self_>().template swizzle<__VA_ARGS__>())> \
+  NAME() {                                                                     \
+    return static_cast<Self_ *>(this)->template swizzle<__VA_ARGS__>();        \
+  }
+
+#define __SYCL_SWIZZLE_MIXIN_METHOD_CONST(COND, NAME, ...)                     \
+  template <int N = NumElements, typename Self_ = Self>                        \
+  std::enable_if_t<(COND), decltype(std::declval<const Self_>()                \
+                                        .template swizzle<__VA_ARGS__>())>     \
+  NAME() const {                                                               \
+    return static_cast<const Self_ *>(this)->template swizzle<__VA_ARGS__>();  \
+  }
+
+#define __SYCL_SWIZLLE_MIXIN_SCALAR_ACCESS_NON_CONST(COND, NAME, INDEX)        \
+  template <int N = NumElements, typename Self_ = Self>                        \
+  std::enable_if_t<(COND), decltype(std::declval<Self_>()[0])> NAME() {        \
+    return (*static_cast<Self_ *>(this))[INDEX];                               \
+  }
+#define __SYCL_SWIZLLE_MIXIN_SCALAR_ACCESS_CONST(COND, NAME, INDEX)            \
+  template <int N = NumElements, typename Self_ = Self>                        \
+  std::enable_if_t<(COND), decltype(std::declval<const Self_>()[0])> NAME()    \
+      const {                                                                  \
+    return (*static_cast<const Self_ *>(this))[INDEX];                         \
+  }
+
+template <typename Self, int NumElements> struct NamedSwizzlesMixinConst {
+#define __SYCL_SWIZZLE_MIXIN_METHOD(COND, NAME, ...)                           \
+  __SYCL_SWIZZLE_MIXIN_METHOD_CONST(COND, NAME, __VA_ARGS__)
+
+#define __SYCL_SWIZLLE_MIXIN_SCALAR_ACCESS(COND, NAME, INDEX)                  \
+  __SYCL_SWIZLLE_MIXIN_SCALAR_ACCESS_CONST(COND, NAME, INDEX)
+
+  __SYCL_SWIZZLE_MIXIN_ALL_SWIZZLES
+
+#undef __SYCL_SWIZLLE_MIXIN_SCALAR_ACCESS
+#undef __SYCL_SWIZZLE_MIXIN_METHOD
+};
+
+template <typename Self, int NumElements> struct NamedSwizzlesMixinBoth {
+#define __SYCL_SWIZZLE_MIXIN_METHOD(COND, NAME, ...)                           \
+  __SYCL_SWIZZLE_MIXIN_METHOD_NON_CONST(COND, NAME, __VA_ARGS__)               \
+  __SYCL_SWIZZLE_MIXIN_METHOD_CONST(COND, NAME, __VA_ARGS__)
+
+#define __SYCL_SWIZLLE_MIXIN_SCALAR_ACCESS(COND, NAME, INDEX)                  \
+  __SYCL_SWIZLLE_MIXIN_SCALAR_ACCESS_NON_CONST(COND, NAME, INDEX)              \
+  __SYCL_SWIZLLE_MIXIN_SCALAR_ACCESS_CONST(COND, NAME, INDEX)
+
+  __SYCL_SWIZZLE_MIXIN_ALL_SWIZZLES
+
+#undef __SYCL_SWIZLLE_MIXIN_SCALAR_ACCESS
+#undef __SYCL_SWIZZLE_MIXIN_METHOD
+};
+
+#undef __SYCL_SWIZZLE_MIXIN_METHOD_CONST
+#undef __SYCL_SWIZZLE_MIXIN_METHOD_NON_CONST
+
+#undef __SYCL_SWIZZLE_MIXIN_SIMPLE_SWIZZLES
```

Comment:
would you like to undef all swizzles also?

Reply:
Yes, good catch, thanks!